### DISCUSSION TITLE
Expanding generic type arguments

### DISCUSSION THREAD

**Discussion by fde-lmag** [dotnet/csharplang#7265] [Language Ideas] [answered]
The idea is to have a syntax to extract generic type arguments from generic type parameters.

Lets assume a generic class
```CSharp
public class Generic<TParam1, TParam2> { ... }
```
and two other classes
```CSharp
public class TypeParameter1 { ... }
public class TypeParameter2 { ... }
```
and now lets create a simple extension method with a type constraint to `Generic`
```CSharp
public static void ExtensionMethod<TGen>(this TGen genericType)
    where TGen : Generic<> { ... }
```
this is obviously incorrect code, it wont compile for good reasons.

But DI has been here for long and we all love it. However, writing an extension Method that looks like this
```CSharp
public static TGen Register<TGen, TParam1, TParam2>(this TGen generic)
    where TGen : Generic<TParam1, TParam2> { ... }
```
and needs to be called like this
```CSharp
var g = new Generic<TypeParameter1, TypeParameter2>();
g.Register<Generic<TypeParameter1, TypeParameter2>, TypeParameter1, TypeParameter2>();
```
is not very likable.

What i want to propose is to have a "generic expansion" feature in the language
```CSharp
public static TGen Register<TGen>(this TGen generic)
    where TGen : Generic<TGen:P1, TGen:P2>
    where P1 : TypeParameter1
    where P2 : TypeParameter2 { ... }
```
and then calling the mthod turns into
```CSharp
g.Register();
```
The Extension method is instructed to extract the generic arguments of TGen into P1 and P2.

Ofc this should also work with derived types
```CSharp
public class GenericDerived : Generic<TypeParameter1, TypeParameter2> { ... }
```
using the extension method would change from
```CSharp
var g = new GeneriDerived();
g.Register<GenericDerived, TypeParameter1, TypeParameter2>();
```
to this
```CSharp
var g = new GenericDerived();
g.Register<GenericDerived>();
// or with deferred Type just
g.Register();
```

I think something like this could really open up extension methods further
as far as i know, if you needed to do something like this today, you would have to use reflection to get the generic type arguments and then reflection to create a generic method.
If this could be a language feature or maybe a compiler feature
it would increase performance and open new possibilities for developers
Especially but not only when it comes to Dependency Injection.

Maybe this is more of a compiler feature than a language feature
but i have to start somewhere ;-)

**Comment by 333fred** [comment, marked as answer]
You want either https://github.com/dotnet/csharplang/issues/5556 or https://github.com/dotnet/csharplang/issues/339.

**Reply by fde-lmag** [reply]
i dont think i really want those

both of those proposals would solve the same sort of "problem" that i describe
i did find both of them while searching
but i believe, they are trying to add way more complex language features

i am no expert in language definitions and compilers
if these two are describing what would be needed under the hood
then i guess i can stop here

but i thought that these proposals go way and beyond
and have a different agenda besides this little "problem" i am describing here
and that extending the type constraints to support "expanding" type arguments
could solve this particular "problem" more easily
more easy to understand when using the language
and more easy to implenent as as feature

(and maybe it would even solve some of the "example problems" of they're proposals probably not all)

from my point of view
i am asking for something different
i am focussing on type constraints
maybe its not so different under the hood
then by all means
disregard

**Reply by 333fred** [reply]
To put it another way: the `Generic` type that you are describing here does not need to be _externally_ generic. For any given implementation, the type implementing it knows exactly what the generic type involved _must_ be, and no other types are allowed. That's exactly the scenario that associated types covers.

**Reply by fde-lmag** [reply]
i dont exactly know what you mean with "externally generic"
But i will try to elaborate a bit further
Lets take my previous example of the `GenericDerived` class
and lets write an extension method where we would like to construct a different generic type with the type arguments that are abstracted away by the specific implementation of `GenericDerived`
So `GenericDerived` is a specification of `Generic<TypeParameter1, TypeParameter2>` and our Extension method (doesnt have to be an extension method) wants to construct a Lambda Function of Func<TypeParameter1, TypeParameter2, bool>
To do that, reflection is needed to get the Type Arguments of the base type of the `GenericDerived` type if we dont want to add the generic method with the additional type arguments of `Generic<TParam1, TParam2>`
What i would like to do would be to just "extract" or "expand" these two type arguments to eliminate the need for Reflection or the need for the additional type arguments which could be inferred from the type and constraint

Today that method would look like this
```CSharp
public static void Register<T, TParam1, TParam2>(this T generic)
    where T : Generic<TParam1, TParam2>
{
    Func<TParam1, TParam2, bool> expression = ...
}
```

but it would be easier and less overhead (since we could eliminate reflection) if we could instead 
```CSharp
public static void Register<T>(this T generic)
    where T : Generic<T:P1, T:P2>
{
    Func<T:P1, T:P2, bool> expression = ...
}
```

In other words, we would define a generic method constrained to an open generic type and would have the possibility to still work with the type arguments that the instance of that open generic type would provide

From what i understand, associated types is more of a filter for specific implementations on a generic implementation where the type arguments would need more precise implementations

**Reply by 333fred** [reply]
> So GenericDerived is a specification of Generic<TypeParameter1, TypeParameter2>

This is the key point. For any given implementation of Generic, the type parameters are implied. There's no need to manually flow them through the type system:  just specifying GenericDerived is enough. While we haven't come up with a syntax for the extraction bit yet, it doesn't seem hard to implement: the compiler would make associated types by readding those extra generic parameters to the method for you, just without you having to manually type them. So a syntax to refer to them seems perfectly reasonable.

**Reply by fde-lmag** [reply]
thx for your explanations and patience ;-)

**Comment by hez2010** [comment]
This is exactly covered by Generics on Generics which has been proposed as #339.

**Reply by fde-lmag** [reply]
if that's the case, then great
